llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeAttention.cpp

```diff
@@ -0,0 +1,397 @@
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/Utils/Utils.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/Support/Debug.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace LinalgExt {
+
+namespace {
+
+// Computes a reduction along the rows of a 2d tensor of shape MxN
+// to produce a tensor of shape M
+template <typename T>
+static Value computeRowwiseReduction(Value a, Value output, Location loc,
+                                     OpBuilder &builder) {
+  SmallVector<utils::IteratorType> iteratorTypes{
+      utils::IteratorType::parallel, utils::IteratorType::reduction};
+  AffineMap id = AffineMap::getMultiDimIdentityMap(2, builder.getContext());
+  AffineExpr d0, d1;
+  bindDims(builder.getContext(), d0, d1);
+  // (d0, d1) -> (d0)
+  auto rowMap = AffineMap::get(2, 0, {d0}, builder.getContext());
+  SmallVector<AffineMap> indexingMaps{id, rowMap};
+  auto genericOp = builder.create<linalg::GenericOp>(
+      loc, output.getType(), a, output, indexingMaps, iteratorTypes,
+      [&](OpBuilder &b, Location loc, ValueRange args) {
+        Value result = b.create<T>(loc, args[0], args[1]);
+        b.create<linalg::YieldOp>(loc, result);
+      });
+  return genericOp.getResult(0);
+}
+
+static std::tuple<Value, Value> computeNewSum(Value oldMax, Value newMax,
+                                              Value oldSum, Value currentSum,
+                                              Value output, Location loc,
+                                              OpBuilder &builder) {
+  SmallVector<utils::IteratorType> iteratorTypes{utils::IteratorType::parallel};
+  auto identityMap = AffineMap::getMultiDimIdentityMap(1, builder.getContext());
+  SmallVector<AffineMap> indexingMaps(6, identityMap);
+  SmallVector<Type> resultTypes(2, currentSum.getType());
+  auto genericOp = builder.create<linalg::GenericOp>(
+      loc, resultTypes, ValueRange{oldMax, newMax, oldSum, currentSum},
+      ValueRange{output, output}, indexingMaps, iteratorTypes,
+      [&](OpBuilder &b, Location loc, ValueRange args) {
+        Value diff = b.create<arith::SubFOp>(loc, args[0], args[1]);
+        Value weight = b.create<math::ExpOp>(loc, diff);
+        Value scaledOldSum = b.create<arith::MulFOp>(loc, weight, args[2]);
+        Value result = b.create<arith::AddFOp>(loc, scaledOldSum, args[3]);
+        b.create<linalg::YieldOp>(loc, ValueRange{result, scaledOldSum});
+      });
+  return std::make_tuple(genericOp.getResult(0), genericOp.getResult(1));
+}
+
+static Value computeSoftmax(Value qkTranspose, Value currentMax, Value output,
+                            Location loc, OpBuilder &builder) {
+  AffineMap identityMap =
```

Comment:
Cant we use the `iree_linalg_ext.softmax` here?

Reply:
This computes a partial softmax, so just exp(x - max(x)). If you would like, I can rename this function?

Reply:
yeah renaming would be good.... Thats a bummer, you had to do all the work for adding softmax op, but wasnt really able to use it here.

Reply:
I know but maybe in the future we can figure out how to leverage that here.